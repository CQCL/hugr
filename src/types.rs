//! General wire types used in the compiler

pub mod custom;
pub mod simple;
pub mod type_param;
pub mod type_row;

use std::fmt::{self, Display, Write};
use std::ops::Index;

#[cfg(feature = "pyo3")]
use pyo3::prelude::*;

pub use custom::CustomType;
pub use simple::{
    ClassicRow, ClassicType, Container, HashableType, PrimType, SimpleRow, SimpleType, TypeTag,
};
pub use type_row::TypeRow;

use delegate::delegate;
use smol_str::SmolStr;

use crate::hugr::{Direction, Port};
use crate::utils::display_list;
use crate::{resource::ResourceSet, type_row};

/// The kinds of edges in a HUGR, excluding Hierarchy.
//#[cfg_attr(feature = "pyo3", pyclass)] # TODO: Manually derive pyclass with non-unit variants
#[derive(Clone, PartialEq, Eq, Debug, serde::Serialize, serde::Deserialize)]
#[non_exhaustive]
pub enum EdgeKind {
    /// Control edges of a CFG region.
    ControlFlow,
    /// Data edges of a DDG region, also known as "wires".
    Value(SimpleType),
    /// A reference to a static value definition.
    Static(ClassicType),
    /// Explicitly enforce an ordering between nodes in a DDG.
    StateOrder,
}

impl EdgeKind {
    /// Returns whether the type might contain linear data.
    pub fn is_linear(&self) -> bool {
        match self {
            EdgeKind::Value(t) => !t.tag().is_classical(),
            _ => false,
        }
    }
}

#[cfg_attr(feature = "pyo3", pyclass)]
#[derive(Clone, Debug, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
/// Describes the edges required to/from a node. This includes both the concept of "signature" in the spec,
/// and also the target (value) of a call (static).
pub struct AbstractSignature {
    /// Value inputs of the function.
    pub input: SimpleRow,
    /// Value outputs of the function.
    pub output: SimpleRow,
    /// Possible static input (for call / load-constant).
    pub static_input: ClassicRow,
    /// The resource requirements which are added by the operation
    pub resource_reqs: ResourceSet,
}

#[derive(Clone, Default, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
/// A concrete signature, which has been instantiated with a set of input resources
pub struct Signature {
    /// The underlying signature
    pub signature: AbstractSignature,
    /// The resources which are associated with all the inputs and carried through
    pub input_resources: ResourceSet,
}

impl AbstractSignature {
    /// Create a new signature.
    pub fn new(
        input: impl Into<SimpleRow>,
        output: impl Into<SimpleRow>,
        static_input: impl Into<ClassicRow>,
    ) -> Self {
        Self {
            input: input.into(),
            output: output.into(),
            static_input: static_input.into(),
            resource_reqs: ResourceSet::new(),
        }
    }

    /// Builder method, add resource_reqs to an AbstractSignature
    pub fn with_resource_delta(mut self, rs: &ResourceSet) -> Self {
        self.resource_reqs = self.resource_reqs.union(rs);
        self
    }

    /// Instantiate an AbstractSignature, converting it to a concrete one
    pub fn with_input_resources(self, rs: ResourceSet) -> Signature {
        Signature {
            signature: self,
            input_resources: rs,
        }
    }

    /// Instantiate a signature with the empty set of resources
    pub fn pure(self) -> Signature {
        self.with_input_resources(ResourceSet::new())
    }
}

impl From<Signature> for AbstractSignature {
    fn from(sig: Signature) -> Self {
        sig.signature
    }
}

impl Signature {
    /// Calculate the resource requirements of the output wires
    pub fn output_resources(&self) -> ResourceSet {
        self.input_resources
            .clone()
            .union(&self.signature.resource_reqs)
    }
}

#[cfg_attr(feature = "pyo3", pymethods)]
impl AbstractSignature {
    /// The number of wires in the signature.
    #[inline(always)]
    pub fn is_empty(&self) -> bool {
        self.static_input.is_empty() && self.input.is_empty() && self.output.is_empty()
    }
}

impl AbstractSignature {
    /// Create a new signature with only dataflow inputs and outputs.
    pub fn new_df(input: impl Into<SimpleRow>, output: impl Into<SimpleRow>) -> Self {
        Self::new(input, output, type_row![])
    }
    /// Create a new signature with the same input and output types.
    pub fn new_linear(linear: impl Into<SimpleRow>) -> Self {
        let linear = linear.into();
        Self::new_df(linear.clone(), linear)
    }

    /// Returns the type of a [`Port`]. Returns `None` if the port is out of bounds.
    pub fn get(&self, port: Port) -> Option<EdgeKind> {
        if port.direction() == Direction::Incoming && port.index() >= self.input.len() {
            Some(EdgeKind::Static(
                self.static_input
                    .get(port.index() - self.input.len())?
                    .clone(),
            ))
        } else {
            self.get_df(port).cloned().map(EdgeKind::Value)
        }
    }

    /// Returns the type of a value [`Port`]. Returns `None` if the port is out
    /// of bounds or if it is not a value.
    #[inline]
    pub fn get_df(&self, port: Port) -> Option<&SimpleType> {
        match port.direction() {
            Direction::Incoming => self.input.get(port.index()),
            Direction::Outgoing => self.output.get(port.index()),
        }
    }

    /// Returns the type of a value [`Port`]. Returns `None` if the port is out
    /// of bounds or if it is not a value.
    #[inline]
    pub fn get_df_mut(&mut self, port: Port) -> Option<&mut SimpleType> {
        match port.direction() {
            Direction::Incoming => self.input.get_mut(port.index()),
            Direction::Outgoing => self.output.get_mut(port.index()),
        }
    }

    /// Returns the number of value and static ports in the signature.
    #[inline]
    pub fn port_count(&self, dir: Direction) -> usize {
        match dir {
            Direction::Incoming => self.input.len() + self.static_input.len(),
            Direction::Outgoing => self.output.len(),
        }
    }

    /// Returns the number of input value and static ports in the signature.
    #[inline]
    pub fn input_count(&self) -> usize {
        self.port_count(Direction::Incoming)
    }

    /// Returns the number of output value and static ports in the signature.
    #[inline]
    pub fn output_count(&self) -> usize {
        self.port_count(Direction::Outgoing)
    }

    /// Returns the number of value ports in the signature.
    #[inline]
    pub fn df_port_count(&self, dir: Direction) -> usize {
        match dir {
            Direction::Incoming => self.input.len(),
            Direction::Outgoing => self.output.len(),
        }
    }

    /// Returns a slice of the value types for the given direction.
    #[inline]
    pub fn df_types(&self, dir: Direction) -> &[SimpleType] {
        match dir {
            Direction::Incoming => &self.input,
            Direction::Outgoing => &self.output,
        }
    }

    /// Returns a slice of the input value types.
    #[inline]
    pub fn input_df_types(&self) -> &[SimpleType] {
        self.df_types(Direction::Incoming)
    }

    /// Returns a slice of the output value types.
    #[inline]
    pub fn output_df_types(&self) -> &[SimpleType] {
        self.df_types(Direction::Outgoing)
    }

    #[inline]
    /// Returns the input row
    pub fn input(&self) -> &SimpleRow {
        &self.input
    }

    #[inline]
    /// Returns the output row
    pub fn output(&self) -> &SimpleRow {
        &self.output
    }

    #[inline]
    /// Returns the row of static inputs
    pub fn static_input(&self) -> &ClassicRow {
        &self.static_input
    }
}

impl AbstractSignature {
    /// Returns the linear part of the signature
    /// TODO: This fails when mixing different linear types.
    #[inline(always)]
    pub fn linear(&self) -> impl Iterator<Item = &SimpleType> {
        debug_assert_eq!(
            self.input
                .iter()
                .filter(|t| !t.tag().is_classical())
                .collect::<Vec<_>>(),
            self.output
                .iter()
                .filter(|t| !t.tag().is_classical())
                .collect::<Vec<_>>()
        );
        self.input.iter().filter(|t| !t.tag().is_classical())
    }

    /// Returns the value `Port`s in the signature for a given direction.
    #[inline]
    pub fn ports_df(&self, dir: Direction) -> impl Iterator<Item = Port> {
        (0..self.df_port_count(dir)).map(move |i| Port::new(dir, i))
    }

    /// Returns the incoming value `Port`s in the signature.
    #[inline]
    pub fn input_ports_df(&self) -> impl Iterator<Item = Port> {
        self.ports_df(Direction::Incoming)
    }

    /// Returns the outgoing value `Port`s in the signature.
    #[inline]
    pub fn output_ports_df(&self) -> impl Iterator<Item = Port> {
        self.ports_df(Direction::Outgoing)
    }

    /// Returns the `Port`s in the signature for a given direction.
    #[inline]
    pub fn ports(&self, dir: Direction) -> impl Iterator<Item = Port> {
        (0..self.port_count(dir)).map(move |i| Port::new(dir, i))
    }

    /// Returns the incoming `Port`s in the signature.
    #[inline]
    pub fn input_ports(&self) -> impl Iterator<Item = Port> {
        self.ports(Direction::Incoming)
    }

    /// Returns the outgoing `Port`s in the signature.
    #[inline]
    pub fn output_ports(&self) -> impl Iterator<Item = Port> {
        self.ports(Direction::Outgoing)
    }
}

impl Signature {
    /// Returns a reference to the resource set for the ports of the
    /// signature in a given direction
    pub fn get_resources(&self, dir: &Direction) -> ResourceSet {
        match dir {
            Direction::Incoming => self.input_resources.clone(),
            Direction::Outgoing => self.output_resources(),
        }
    }

    delegate! {
        to self.signature {
            /// Inputs of the abstract signature
            pub fn input(&self) -> &SimpleRow;
            /// Outputs of the abstract signature
            pub fn output(&self) -> &SimpleRow;
            /// Static inputs of the abstract signature
            pub fn static_input(&self) -> &ClassicRow;
        }
    }
}

impl Display for AbstractSignature {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let has_inputs = !(self.static_input.is_empty() && self.input.is_empty());
        if has_inputs {
            self.input.fmt(f)?;
            if !self.static_input.is_empty() {
                f.write_char('<')?;
                display_list(&self.static_input, f)?;
                f.write_char('>')?;
            }
            f.write_str(" -> ")?;
        }
        f.write_char('[')?;
        self.resource_reqs.fmt(f)?;
        f.write_char(']')?;
        self.output.fmt(f)
    }
}

impl Display for Signature {
    delegate! {
        to self.signature {
            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result;
        }
    }
}

/// Descriptive names for the ports in a [`Signature`].
///
/// This is a separate type from [`Signature`] as it is not normally used during the compiler operations.
#[cfg_attr(feature = "pyo3", pyclass)]
#[derive(Clone, Default, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct SignatureDescription {
    /// Input of the function.
    pub input: Vec<SmolStr>,
    /// Output of the function.
    pub output: Vec<SmolStr>,
    /// Static data references used by the function.
    pub static_input: Vec<SmolStr>,
}

#[cfg_attr(feature = "pyo3", pymethods)]
impl SignatureDescription {
    /// The number of wires in the signature.
    #[inline(always)]
    pub fn is_empty(&self) -> bool {
        self.static_input.is_empty() && self.input.is_empty() && self.output.is_empty()
    }
}

impl SignatureDescription {
    /// Create a new signature.
    pub fn new(
        input: impl Into<Vec<SmolStr>>,
        output: impl Into<Vec<SmolStr>>,
        static_input: impl Into<Vec<SmolStr>>,
    ) -> Self {
        Self {
            input: input.into(),
            output: output.into(),
            static_input: static_input.into(),
        }
    }

    /// Create a new signature with only linear dataflow inputs and outputs.
    pub fn new_linear(linear: impl Into<Vec<SmolStr>>) -> Self {
        let linear = linear.into();
        SignatureDescription::new_df(linear.clone(), linear)
    }

    /// Create a new signature with only dataflow inputs and outputs.
    pub fn new_df(input: impl Into<Vec<SmolStr>>, output: impl Into<Vec<SmolStr>>) -> Self {
        Self {
            input: input.into(),
            output: output.into(),
            ..Default::default()
        }
    }

    fn row_zip<'a, T: PrimType>(
        type_row: &'a TypeRow<T>,
        name_row: &'a [SmolStr],
    ) -> impl Iterator<Item = (&'a SmolStr, &'a T)> {
        name_row
            .iter()
            .chain(&EmptyStringIterator)
            .zip(type_row.iter())
    }

    /// Iterate over the input wires of the signature and their names.
    ///
    /// Unnamed wires are given an empty string name.
    ///
    /// TODO: Return Option<&String> instead of &String for the description.
    pub fn input_zip<'a>(
        &'a self,
        signature: &'a Signature,
    ) -> impl Iterator<Item = (&SmolStr, &SimpleType)> {
        Self::row_zip(signature.input(), &self.input)
    }

    /// Iterate over the output wires of the signature and their names.
    ///
    /// Unnamed wires are given an empty string name.
    pub fn output_zip<'a>(
        &'a self,
        signature: &'a Signature,
    ) -> impl Iterator<Item = (&SmolStr, &SimpleType)> {
        Self::row_zip(signature.output(), &self.output)
    }

    /// Iterate over the static input wires of the signature and their names.
    pub fn static_input_zip<'a>(
        &'a self,
        signature: &'a Signature,
    ) -> impl Iterator<Item = (&SmolStr, &ClassicType)> {
        Self::row_zip(signature.static_input(), &self.static_input)
    }
}

impl Index<Port> for SignatureDescription {
    type Output = SmolStr;

    fn index(&self, index: Port) -> &Self::Output {
        match index.direction() {
            Direction::Incoming => self.input.get(index.index()).unwrap_or(EMPTY_STRING_REF),
            Direction::Outgoing => self.output.get(index.index()).unwrap_or(EMPTY_STRING_REF),
        }
    }
}

/// An iterator that always returns the an empty string.
struct EmptyStringIterator;

/// A reference to an empty string. Used by [`EmptyStringIterator`].
const EMPTY_STRING_REF: &SmolStr = &SmolStr::new_inline("");

impl<'a> Iterator for &'a EmptyStringIterator {
    type Item = &'a SmolStr;

    fn next(&mut self) -> Option<Self::Item> {
        Some(EMPTY_STRING_REF)
    }
}
