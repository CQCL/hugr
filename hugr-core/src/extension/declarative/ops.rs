//! Declarative operation definitions.
//!
//! This module defines a YAML schema for defining operations in a declarative way.
//!
//! See the [specification] and [`ExtensionSetDeclaration`] for more details.
//!
//! [specification]: https://github.com/CQCL/hugr/blob/main/specification/hugr.md#declarative-format
//! [`ExtensionSetDeclaration`]: super::ExtensionSetDeclaration

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use smol_str::SmolStr;

use crate::extension::{OpDef, SignatureFunc};
use crate::ops::OpName;
use crate::types::type_param::TypeParam;
use crate::Extension;

use super::signature::SignatureDeclaration;
use super::{DeclarationContext, ExtensionDeclarationError};

/// A declarative operation definition.
///
/// TODO: The "Lowering" attribute is not yet supported.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(super) struct OperationDeclaration {
    /// The identifier the operation.
    name: OpName,
    /// A description for the operation.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    description: String,
    /// The signature of the operation.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    signature: Option<SignatureDeclaration>,
    /// A set of per-node parameters required to instantiate this operation.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    params: HashMap<SmolStr, ParamDeclaration>,
    /// An extra set of data associated to the operation.
    ///
    /// This data is kept in the Hugr, and may be accessed by the relevant runtime.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    misc: HashMap<String, serde_yaml::Value>,
    /// A pre-compiled lowering routine.
    ///
    /// This is not yet supported, and will raise an error if present.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    lowering: Option<String>,
}

impl OperationDeclaration {
    /// Register this operation in the given extension.
    pub fn register<'ext>(
        &self,
        ext: &'ext mut Extension,
        ctx: DeclarationContext<'_>,
    ) -> Result<&'ext mut OpDef, ExtensionDeclarationError> {
        // We currently only support explicit signatures.
        //
        // TODO: Support missing signatures?
        let Some(signature) = &self.signature else {
            return Err(ExtensionDeclarationError::MissingSignature {
                ext: ext.name().clone(),
                op: self.name.clone(),
            });
        };

        // We currently do not support parametric operations.
        if !self.params.is_empty() {
            return Err(ExtensionDeclarationError::ParametricOperation {
                ext: ext.name().clone(),
                op: self.name.clone(),
            });
        }
        let params: Vec<TypeParam> = vec![];

        if self.lowering.is_some() {
            return Err(ExtensionDeclarationError::LoweringNotSupported {
                ext: ext.name().clone(),
                op: self.name.clone(),
            });
        }

        let signature_func: SignatureFunc = signature.make_signature(ext, ctx, &params)?;

        let op_def = ext.add_op(self.name.clone(), self.description.clone(), signature_func)?;

        for (k, v) in &self.misc {
            op_def.add_misc(k, v.clone());
        }

        Ok(op_def)
    }
}

/// The type of a per-node operation parameter required to instantiate an operation.
///
/// TODO: The value should be decoded as a [`TypeParam`].
/// Valid options include:
///
/// - `USize`
/// - `Type`
///
/// [`TypeParam`]: crate::types::type_param::TypeParam
#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub(super) struct ParamDeclaration(
    /// TODO: Store a [`TypeParam`], and implement custom parsers.
    ///
    /// [`TypeParam`]: crate::types::type_param::TypeParam
    String,
);
