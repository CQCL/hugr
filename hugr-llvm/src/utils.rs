//! Module for utilities that do not depend on LLVM. These are candidates for
//! upstreaming.
#[deprecated(note = "This module is deprecated and will be removed in a future release.")]
pub mod array_op_builder;
pub mod fat;
pub mod inline_constant_functions;
pub mod int_op_builder;
pub mod logic_op_builder;
pub mod type_map;

#[deprecated(note = "Import from hugr_core::std_extensions::collections::array.")]
pub use hugr_core::std_extensions::collections::array::ArrayOpBuilder;
pub use inline_constant_functions::inline_constant_functions;
pub use int_op_builder::IntOpBuilder;
pub use logic_op_builder::LogicOpBuilder;
