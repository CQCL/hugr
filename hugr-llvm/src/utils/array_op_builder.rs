use hugr_core::std_extensions::collections::array::{
    Array, ArrayKind, GenericArrayClone, GenericArrayDiscard, GenericArrayOpDef,
};
use hugr_core::{
    builder::{BuildError, Dataflow},
    extension::simple_op::HasConcrete as _,
    types::Type,
    Wire,
};
use itertools::Itertools as _;

pub trait ArrayOpBuilder<AK: ArrayKind>: Dataflow {
    fn add_new_array(
        &mut self,
        elem_ty: Type,
        values: impl IntoIterator<Item = Wire>,
    ) -> Result<Wire, BuildError> {
        let inputs = values.into_iter().collect_vec();
        let [out] = self
            .add_dataflow_op(
                GenericArrayOpDef::<AK>::new_array.to_concrete(elem_ty, inputs.len() as u64),
                inputs,
            )?
            .outputs_arr();
        Ok(out)
    }

    fn add_array_clone(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
    ) -> Result<(Wire, Wire), BuildError> {
        let op = GenericArrayClone::<AK>::new(elem_ty, size).unwrap();
        let [arr1, arr2] = self.add_dataflow_op(op, vec![input])?.outputs_arr();
        Ok((arr1, arr2))
    }

    fn add_array_discard(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
    ) -> Result<(), BuildError> {
        let op = GenericArrayDiscard::<AK>::new(elem_ty, size).unwrap();
        let [] = self.add_dataflow_op(op, vec![input])?.outputs_arr();
        Ok(())
    }

    fn add_array_get(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
        index: Wire,
    ) -> Result<(Wire, Wire), BuildError> {
        // TODO Add an OpLoadError variant to BuildError.
        let op = GenericArrayOpDef::<AK>::get
            .instantiate(&[size.into(), elem_ty.into()])
            .unwrap();
        let [out, arr] = self.add_dataflow_op(op, vec![input, index])?.outputs_arr();
        Ok((out, arr))
    }

    fn add_array_set(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
        index: Wire,
        value: Wire,
    ) -> Result<Wire, BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        let op = GenericArrayOpDef::<AK>::set
            .instantiate(&[size.into(), elem_ty.into()])
            .unwrap();
        let [out] = self
            .add_dataflow_op(op, vec![input, index, value])?
            .outputs_arr();
        Ok(out)
    }

    fn add_array_swap(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
        index1: Wire,
        index2: Wire,
    ) -> Result<Wire, BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        let op = GenericArrayOpDef::<AK>::swap
            .instantiate(&[size.into(), elem_ty.into()])
            .unwrap();
        let [out] = self
            .add_dataflow_op(op, vec![input, index1, index2])?
            .outputs_arr();
        Ok(out)
    }

    fn add_array_pop_left(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
    ) -> Result<Wire, BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        let op = GenericArrayOpDef::<AK>::pop_left
            .instantiate(&[size.into(), elem_ty.into()])
            .unwrap();
        Ok(self.add_dataflow_op(op, vec![input])?.out_wire(0))
    }

    fn add_array_pop_right(
        &mut self,
        elem_ty: Type,
        size: u64,
        input: Wire,
    ) -> Result<Wire, BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        let op = GenericArrayOpDef::<AK>::pop_right
            .instantiate(&[size.into(), elem_ty.into()])
            .unwrap();
        Ok(self.add_dataflow_op(op, vec![input])?.out_wire(0))
    }

    fn add_array_discard_empty(&mut self, elem_ty: Type, input: Wire) -> Result<(), BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        self.add_dataflow_op(
            GenericArrayOpDef::<AK>::discard_empty
                .instantiate(&[elem_ty.into()])
                .unwrap(),
            [input],
        )?;
        Ok(())
    }
}

impl<D: Dataflow> ArrayOpBuilder<Array> for D {}

#[cfg(test)]
pub mod test {
    use hugr_core::extension::prelude::PRELUDE_ID;
    use hugr_core::extension::ExtensionSet;
    use hugr_core::std_extensions::collections::array::{self, array_type};
    use hugr_core::{
        builder::{DFGBuilder, HugrBuilder},
        extension::prelude::{either_type, option_type, usize_t, ConstUsize, UnwrapBuilder as _},
        types::Signature,
        Hugr,
    };
    use rstest::rstest;

    use super::*;

    #[rstest::fixture]
    #[default(DFGBuilder<Hugr>)]
    pub fn all_array_ops<B: Dataflow>(
        #[default(DFGBuilder::new(Signature::new_endo(Type::EMPTY_TYPEROW)
            .with_extension_delta(ExtensionSet::from_iter([
                PRELUDE_ID,
                array::EXTENSION_ID
        ]))).unwrap())]
        mut builder: B,
    ) -> B {
        let us0 = builder.add_load_value(ConstUsize::new(0));
        let us1 = builder.add_load_value(ConstUsize::new(1));
        let us2 = builder.add_load_value(ConstUsize::new(2));
        let arr = builder.add_new_array(usize_t(), [us1, us2]).unwrap();
        let [arr] = {
            let r = builder.add_array_swap(usize_t(), 2, arr, us0, us1).unwrap();
            let res_sum_ty = {
                let array_type = array_type(2, usize_t());
                either_type(array_type.clone(), array_type)
            };
            builder.build_unwrap_sum(1, res_sum_ty, r).unwrap()
        };

        let ([elem_0], arr) = {
            let (r, arr) = builder.add_array_get(usize_t(), 2, arr, us0).unwrap();
            (
                builder
                    .build_unwrap_sum(1, option_type(usize_t()), r)
                    .unwrap(),
                arr,
            )
        };

        let [_elem_1, arr] = {
            let r = builder
                .add_array_set(usize_t(), 2, arr, us1, elem_0)
                .unwrap();
            let res_sum_ty = {
                let row = vec![usize_t(), array_type(2, usize_t())];
                either_type(row.clone(), row)
            };
            builder.build_unwrap_sum(1, res_sum_ty, r).unwrap()
        };

        let [_elem_left, arr] = {
            let r = builder.add_array_pop_left(usize_t(), 2, arr).unwrap();
            builder
                .build_unwrap_sum(1, option_type(vec![usize_t(), array_type(1, usize_t())]), r)
                .unwrap()
        };
        let [_elem_right, arr] = {
            let r = builder.add_array_pop_right(usize_t(), 1, arr).unwrap();
            builder
                .build_unwrap_sum(1, option_type(vec![usize_t(), array_type(0, usize_t())]), r)
                .unwrap()
        };

        builder.add_array_discard_empty(usize_t(), arr).unwrap();
        builder
    }

    #[rstest]
    fn build_all_ops(all_array_ops: DFGBuilder<Hugr>) {
        all_array_ops.finish_hugr().unwrap();
    }
}
