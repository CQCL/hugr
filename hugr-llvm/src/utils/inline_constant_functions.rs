use hugr::{
    extension::ExtensionRegistry,
    hugr::hugrmut::HugrMut,
    ops::{FuncDefn, LoadFunction, Value},
    types::PolyFuncType,
    HugrView, Node, NodeIndex as _,
};

use anyhow::{anyhow, bail, Result};

fn const_fn_name(konst_n: Node) -> String {
    format!("const_fun_{}", konst_n.index())
}

pub fn inline_constant_functions(
    hugr: &mut impl HugrMut,
    registry: &ExtensionRegistry,
) -> Result<()> {
    while inline_constant_functions_impl(hugr, registry)? {}
    Ok(())
}

fn inline_constant_functions_impl(
    hugr: &mut impl HugrMut,
    registry: &ExtensionRegistry,
) -> Result<bool> {
    let mut const_funs = vec![];

    for n in hugr.nodes() {
        let konst_hugr = {
            let Some(konst) = hugr.get_optype(n).as_const() else {
                continue;
            };
            let Value::Function { hugr } = konst.value() else {
                continue;
            };
            let optype = hugr.get_optype(hugr.root());
            if !optype.is_dfg() && !optype.is_func_defn() {
                bail!(
                    "Constant function has unsupported root: {:?}",
                    hugr.get_optype(hugr.root())
                )
            }
            hugr.clone()
        };
        let mut lcs = vec![];
        for load_constant in hugr.output_neighbours(n) {
            if !hugr.get_optype(load_constant).is_load_constant() {
                bail!(
                    "Constant function has non-LoadConstant output-neighbour: {load_constant} {:?}",
                    hugr.get_optype(load_constant)
                )
            }
            lcs.push(load_constant);
        }
        const_funs.push((n, konst_hugr.as_ref().clone(), lcs));
    }

    let mut any_changes = false;

    for (konst_n, func_hugr, load_constant_ns) in const_funs {
        if !load_constant_ns.is_empty() {
            let polysignature: PolyFuncType = func_hugr
                .inner_function_type()
                .ok_or(anyhow!(
                    "Constant function hugr has no inner_func_type: {}",
                    konst_n.index()
                ))?
                .into();
            let func_defn = FuncDefn {
                name: const_fn_name(konst_n),
                signature: polysignature.clone(),
            };
            let func_node = hugr.add_node_with_parent(hugr.root(), func_defn);
            hugr.insert_hugr(func_node, func_hugr);

            for lcn in load_constant_ns {
                hugr.replace_op(
                    lcn,
                    LoadFunction::try_new(polysignature.clone(), [], registry)?,
                )?;
            }
            any_changes = true;
        }
        hugr.remove_node(konst_n);
    }
    Ok(any_changes)
}

#[cfg(test)]
mod test {
    use hugr::{
        builder::{
            Container, DFGBuilder, Dataflow, DataflowHugr, DataflowSubContainer, HugrBuilder,
            ModuleBuilder,
        },
        extension::{prelude::QB_T, PRELUDE_REGISTRY},
        ops::{CallIndirect, Const, Value},
        types::Signature,
        Hugr, HugrView, Wire,
    };

    use super::inline_constant_functions;

    fn build_const(go: impl FnOnce(&mut DFGBuilder<Hugr>) -> Wire) -> Const {
        Value::function({
            let mut builder = DFGBuilder::new(Signature::new_endo(QB_T)).unwrap();
            let r = go(&mut builder);
            builder
                .finish_hugr_with_outputs([r], &PRELUDE_REGISTRY)
                .unwrap()
        })
        .unwrap()
        .into()
    }

    #[test]
    fn simple() {
        let qb_sig: Signature = Signature::new_endo(QB_T);
        let mut hugr = {
            let mut builder = ModuleBuilder::new();
            let const_node = builder.add_constant(build_const(|builder| {
                let [r] = builder.input_wires_arr();
                r
            }));
            {
                let mut builder = builder.define_function("main", qb_sig.clone()).unwrap();
                let [i] = builder.input_wires_arr();
                let fun = builder.load_const(&const_node);
                let [r] = builder
                    .add_dataflow_op(
                        CallIndirect {
                            signature: qb_sig.clone(),
                        },
                        [fun, i],
                    )
                    .unwrap()
                    .outputs_arr();
                builder.finish_with_outputs([r]).unwrap();
            };
            builder.finish_hugr(&PRELUDE_REGISTRY).unwrap()
        };

        inline_constant_functions(&mut hugr, &PRELUDE_REGISTRY).unwrap();

        for n in hugr.nodes() {
            if let Some(konst) = hugr.get_optype(n).as_const() {
                assert!(!matches!(konst.value(), Value::Function { .. }))
            }
        }
    }

    #[test]
    fn nested() {
        let qb_sig: Signature = Signature::new_endo(QB_T);
        let mut hugr = {
            let mut builder = ModuleBuilder::new();
            let const_node = builder.add_constant(build_const(|builder| {
                let [i] = builder.input_wires_arr();
                let func = builder.add_load_const(build_const(|builder| {
                    let [r] = builder.input_wires_arr();
                    r
                }));
                let [r] = builder
                    .add_dataflow_op(
                        CallIndirect {
                            signature: qb_sig.clone(),
                        },
                        [func, i],
                    )
                    .unwrap()
                    .outputs_arr();
                r
            }));
            {
                let mut builder = builder.define_function("main", qb_sig.clone()).unwrap();
                let [i] = builder.input_wires_arr();
                let fun = builder.load_const(&const_node);
                let [r] = builder
                    .add_dataflow_op(
                        CallIndirect {
                            signature: qb_sig.clone(),
                        },
                        [fun, i],
                    )
                    .unwrap()
                    .outputs_arr();
                builder.finish_with_outputs([r]).unwrap();
            };
            builder.finish_hugr(&PRELUDE_REGISTRY).unwrap()
        };

        inline_constant_functions(&mut hugr, &PRELUDE_REGISTRY).unwrap();

        for n in hugr.nodes() {
            if let Some(konst) = hugr.get_optype(n).as_const() {
                assert!(!matches!(konst.value(), Value::Function { .. }))
            }
        }
    }
}
