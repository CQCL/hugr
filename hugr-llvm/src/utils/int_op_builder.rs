use hugr::{
    builder::{BuildError, Dataflow},
    extension::simple_op::HasConcrete as _,
    std_extensions::arithmetic::int_ops::IntOpDef,
    types::TypeArg,
    Wire,
};

pub trait IntOpBuilder: Dataflow {
    fn add_iadd(
        &mut self,
        log_width: impl Into<TypeArg>,
        x1: Wire,
        x2: Wire,
    ) -> Result<Wire, BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        let op = IntOpDef::iadd.instantiate(&[log_width.into()]).unwrap();
        Ok(self.add_dataflow_op(op, [x1, x2])?.out_wire(0))
    }

    fn add_ieq(
        &mut self,
        log_width: impl Into<TypeArg>,
        x1: Wire,
        x2: Wire,
    ) -> Result<Wire, BuildError> {
        // TODO Add an OpLoadError variant to BuildError
        let op = IntOpDef::ieq.instantiate(&[log_width.into()]).unwrap();
        Ok(self.add_dataflow_op(op, [x1, x2])?.out_wire(0))
    }
}

impl<D: Dataflow> IntOpBuilder for D {}
