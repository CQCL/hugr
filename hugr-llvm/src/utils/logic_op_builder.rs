use hugr::{
    builder::{BuildError, Dataflow},
    std_extensions::logic::LogicOp,
    Wire,
};

pub trait LogicOpBuilder: Dataflow {
    fn add_and(&mut self, x1: Wire, x2: Wire) -> Result<Wire, BuildError> {
        Ok(self.add_dataflow_op(LogicOp::And, [x1, x2])?.out_wire(0))
    }

    fn add_not(&mut self, x1: Wire) -> Result<Wire, BuildError> {
        Ok(self.add_dataflow_op(LogicOp::Not, [x1])?.out_wire(0))
    }
}

impl<D: Dataflow> LogicOpBuilder for D {}
