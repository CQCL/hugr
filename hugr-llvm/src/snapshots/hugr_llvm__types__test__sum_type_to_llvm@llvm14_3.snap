---
source: src/types.rs
expression: "[]+[]"
---
"{ i32, {}, {} }"
